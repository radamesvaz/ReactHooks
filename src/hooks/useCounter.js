import { useState } from "react"

export const useCounter = ( defaultState = 10 ) => {

    const [counter, setCounter] = useState( defaultState );

    const increment = () => {
        setCounter( prevState => prevState + 1 )
    }

    // const decrement = (factor = 1) => {
    const decrement = () => {
        setCounter( prevState => prevState - 1 )
    }

    const reset = () =>{
        setCounter( defaultState )
    }

    return {
        counter,
        increment,
        decrement,
        reset
    }
}
