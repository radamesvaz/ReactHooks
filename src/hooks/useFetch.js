import { useEffect, useRef, useState } from "react"

export const useFetch = ( url ) => {

    const isMounted = useRef(true);
    
    const [state, setState] = useState({
        data: null,
        loading: true,
        error: null
    })

    useEffect(() => {
        
        return () => {
            isMounted.current = false;
        }
    }, [])

    useEffect( () => {

        setState({
            data: null,
            loading: true,
            error: null
        })

        fetch(url)
            .then(res => res.json())
            .then(data => {
                     isMounted.current && (
                         setState( prevState => ({
                             ...prevState,
                             data,
                             loading: false
                         }) 
                        )
                     )

            })
            .catch( () => {
                setState( prevState => ({
                    data: null,
                    loading: true,
                    error: 'No se pudo buscar la informacion'
                }))
            })
    }, [ url ])

    return state;
}

/*  
    VIDEO: 114 useRef - Caso de uso real

    Varias cosas que anotar en este hook: El useRef nos sirve para mantener la referencia a una variable SIN renderizar de nuevo el componente.
    Ejemplo en el primer useEffect (Recordar Dimedisa y el window.unload) cambiamos el valor de la variable isMounted en la función de cleanup
    y después se comprueba el valor de esa variable cuando se tiene la data del fetch: Línea 31.
    Si todavía el componente esta montado (TRUE) entonces se cambia el estado y se renderiza el componente, de lo contrario no se hace nada y se evitan errores
    al querer renderizar un componente que fue desmontado.

    Importante: la función cleanup debería ir en el segundo useEffect, pero se creó uno nuevo para observar el ejercicio claramente
*/