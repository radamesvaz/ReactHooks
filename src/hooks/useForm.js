import { useState } from "react"

export const useForm = ( initialState = {} ) => {

    const [values, setValues] = useState(initialState);
    // console.log(initialState)

    const reset = () => {
        setValues(initialState)
    }

    const handleInputChange = ({ target }) => {
        setValues(prevValues => ({
            ...prevValues,
            [ target.name ]: target.value
        }))
    } 

    return [values, handleInputChange, reset]
}

// El reset va a devolver el estado a como fue recibido
