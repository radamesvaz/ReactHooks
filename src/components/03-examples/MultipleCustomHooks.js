import React from 'react';
import { useCounter } from '../../hooks/useCounter';
import { useFetch } from '../../hooks/useFetch';
  
import '../02-useEffect/effect.css'

export const MultipleCustomHooks = () => {

    const { counter, increment } = useCounter(1)

    const url = `https://www.breakingbadapi.com/api/quotes/${counter}`

    const  { loading, data }  = useFetch( url );

    const { quote, author } = !!data && data[0];    // Convertimos el null en doble negaci√≥n para evaluar si existe y asi asignar las variables


    return (
        <div>
            <h1>Multiples hooks</h1>
            <hr />

            {/* {
                loading ? <p>Cargando Cita...</p> : <h3>{data[0].quote}</h3>
            } */}

            {
                loading
                    ?
                    (
                        <div className="alert alert-info text-center">
                            Loading...
                        </div>
                    )
                    :
                    (
                        <blockquote className="blockquote text-right">
                            <p className="mb-0"> {quote} </p>
                            <hr />
                            <footer className="blockquote-footer"> {author} </footer>
                        </blockquote>
                    )
            }
            <button 
                className = "btn btn-primary"
                onClick = { increment }
            >
                Siguiente cita
            </button>

        </div>
    )
}
