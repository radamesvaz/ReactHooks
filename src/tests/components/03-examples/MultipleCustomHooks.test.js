import React from 'react';
import { shallow } from 'enzyme';
import "@testing-library/jest-dom";
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';

jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');



describe('Haciendo pruebas sobre el componente <MultipleCustomHooks />', () => {

    useCounter.mockReturnValue({
        counter: 2,
        increment: () => {}
    })
    
    test('debe de mostrar el componente correctamente ', () => {

        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: false
        })
        
        const wrapper = shallow(<MultipleCustomHooks />);

        expect(wrapper).toMatchSnapshot()
    })

    test('debe de mostrar la información conseguida del useFetch ', () => {
        
        useFetch.mockReturnValue({
            data: [{
                author: 'Radames',
                quote: 'Quiero irme de aquí'
            }],
            loading: false,
            error: false
        })

        const wrapper = shallow(<MultipleCustomHooks />);

        expect(wrapper).toMatchSnapshot();

        expect( wrapper.find('.alert-info').exists() ).toBe(false);
        expect( wrapper.find('.mb-0').text().trim() ).toBe('Quiero irme de aquí')

    })
    
    
})
