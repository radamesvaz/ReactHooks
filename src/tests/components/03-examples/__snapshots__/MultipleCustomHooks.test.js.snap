// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Haciendo pruebas sobre el componente <MultipleCustomHooks /> debe de mostrar el componente correctamente  1`] = `
<div>
  <h1>
    Multiples hooks
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente cita
  </button>
</div>
`;

exports[`Haciendo pruebas sobre el componente <MultipleCustomHooks /> debe de mostrar la información conseguida del useFetch  1`] = `
<div>
  <h1>
    Multiples hooks
  </h1>
  <hr />
  <blockquote
    className="blockquote text-right"
  >
    <p
      className="mb-0"
    >
       
      Quiero irme de aquí
       
    </p>
    <hr />
    <footer
      className="blockquote-footer"
    >
       
      Radames
       
    </footer>
  </blockquote>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente cita
  </button>
</div>
`;
